cli/src/main.rs

```diff
@@ -37,172 +54,393 @@ struct Args {
-    } else {
-        match compile_file(&args) {
-            Ok((tree, state)) => {
-                let mut stdout = stdout();
-
-                stdout.execute(terminal::Clear(terminal::ClearType::All))?;
-                stdout.execute(cursor::MoveTo(0, 0))?;
-                stdout.execute(style::PrintStyledContent(
-                    "File successfully compiled!\n\n".green(),
-                ))?;
-
-                if !state.warnings.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Warnings:\n".yellow()))?;
-                    for warning in state.warnings {
-                        stdout.execute(style::PrintStyledContent(format!("{warning}").yellow()))?;
-                    }
-                }
+        // Just start the file watcher, but without html live preview
+        // which means that we wil provide None instead of the document and connections
+        return watch_files(None, None, &args, &current_path).await;
+    }
 
-                if !state.errors.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Errors:\n".red()))?;
-                    for error in state.errors {
-                        stdout.execute(style::PrintStyledContent(format!("{error}").red()))?;
-                    }
-                }
+    // Otherwise, if they are using the watcher or live preview
+    // just compile the file once and save it.
+    print_compiling_message()?;
+    let compilation_result = compile_file(&args.input, &args.get_output_format()?);
+    save_result(&compilation_result, &args)?;
+    print_result(&compilation_result, &args)?;
+
+    Ok(())
+}
+
+/// Choose a free port for hosting the html live preview
+fn get_port() -> Result<u16, CliError> {
+    PREVIEW_PORT
+        .get_or_init(|| portpicker::pick_unused_port())
+        .ok_or(CliError::NoFreePorts)
+}
+
+fn get_server_config(
+    document: PreviewDoc,
+    connections: PreviewConnections,
+    current_path: PathBuf,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone + 'static {
+    let connections = warp::any().map(move || connections.clone());
+    let document = warp::any().map(move || document.clone());
+
+    let websocket = warp::path("ws").and(warp::ws()).and(connections).map(
+        |ws: warp::ws::Ws, connections: PreviewConnections| {
+            ws.on_upgrade(move |socket| handle_connection(socket, connections.clone()))
+        },
+    );
+
+    let preview = warp::path::end().and(document).map(|document: PreviewDoc| {
+        // Inject a JS script after the end of the </body> tg
+        let html = document.lock().unwrap();
+        let modified_html = RE_INJECTION.replace(
+            &html,
+            concat!("</body>\n", include_str!("./preview_injection.html")),
+        );
+        warp::reply::html(modified_html.to_string())
+    });
+
+    let working_directory = warp::fs::dir(current_path);
+
+    let routes = working_directory.or(websocket).or(preview);
```

Comment:
You don't have to bind this to a variable, you can implicitly return it without binding